crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright 2024 Mauro Romito
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types for searching the public room directory.
+
+use eyeball_im::{ObservableVector, VectorDiff};
+use futures_core::Stream;
+use imbl::Vector;
+use ruma::{
+    api::client::directory::get_public_rooms_filtered::v3::Request as PublicRoomsFilterRequest,
+    directory::{Filter, PublicRoomJoinRule},
+    OwnedMxcUri, OwnedRoomAliasId, OwnedRoomId,
+};
+
+use crate::{Client, Result};
+
+/// This struct represents a single result of a room directory search.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RoomDescription {
+    /// The room's ID.
+    pub room_id: OwnedRoomId,
+    /// The name of the room, if any.
+    pub name: Option<String>,
+    /// The topic of the room, if any.
+    pub topic: Option<String>,
+    /// The canonical alias of the room, if any.
+    pub alias: Option<OwnedRoomAliasId>,
+    /// The room's avatar URL, if any.
+    pub avatar_url: Option<OwnedMxcUri>,
+    /// The room's join rule.
+    pub join_rule: PublicRoomJoinRule,
+    /// Whether can be previewed
+    pub is_world_readable: bool,
+    /// The number of members that have joined the room.
+    pub joined_members: u64,
+}
+
+impl From<ruma::directory::PublicRoomsChunk> for RoomDescription {
+    fn from(value: ruma::directory::PublicRoomsChunk) -> Self {
+        Self {
+            room_id: value.room_id,
+            name: value.name,
+            topic: value.topic,
+            alias: value.canonical_alias,
+            avatar_url: value.avatar_url,
+            join_rule: value.join_rule,
+            is_world_readable: value.world_readable,
+            joined_members: value.num_joined_members.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct RoomDirectorySearch {
+    batch_size: u32,
+    filter: Option<String>,
+    next_token: Option<String>,
+    client: Client,
+    results: ObservableVector<RoomDescription>,
+    is_at_last_page: bool,
+}
```

Comment:
Add documentation please :-). Adding an example would be lovely too:

```rust
/// My doc
///
/// # Example
///
/// ```rust
/// use matrix_sdk::room_directory_search::RoomDirectorySearch;
///
/// # fn main() {
/// let room_directory_search = …;
/// # }
/// ```
```
The `# …` trick allows to hide some code from the example in the documentation output, but it's actually used by the example itself when run as a test program.